src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1013 @@
+  if (status.ok() && !shutting_down_) {
+    ares_process_fd(channel_, ARES_SOCKET_BAD, fd_node->as);
+  } else {
+    // If error is not absl::OkStatus() or the resolution was cancelled, it
+    // means the fd has been shutdown or timed out. The pending lookups made
+    // on this ev_driver will be cancelled by the following ares_cancel() and
+    // the on_done callbacks will be invoked with a status of ARES_ECANCELLED.
+    // The remaining file descriptors in this ev_driver will be cleaned up in
+    // the follwing grpc_ares_notify_on_event_locked().
+    ares_cancel(channel_);
+  }
+  Work();
+}
+
+void GrpcAresRequest::OnQueryTimeout() {
+  {
+    grpc_core::MutexLock lock(&mu_);
+    query_timeout_handle_.reset();
+    GRPC_ARES_WRAPPER_TRACE_LOG("request:%p OnQueryTimeout. shutting_down_=%d",
+                                this, shutting_down_);
+    if (!shutting_down_) {
+      shutting_down_ = true;
+      ShutdownPolledFdsLocked(
+          grpc_core::StatusCreate(absl::StatusCode::kDeadlineExceeded,
+                                  "OnQueryTimeout", DEBUG_LOCATION, {}));
+    }
+  }
+  Unref(DEBUG_LOCATION, "OnQueryTimeout");
+}
+
+// In case of non-responsive DNS servers, dropped packets, etc., c-ares has
+// intelligent timeout and retry logic, which we can take advantage of by
+// polling ares_process_fd on time intervals. Overall, the c-ares library is
+// meant to be called into and given a chance to proceed name resolution:
+//   a) when fd events happen
+//   b) when some time has passed without fd events having happened
+// For the latter, we use this backup poller. Also see
+// https://github.com/grpc/grpc/pull/17688 description for more details.
+void GrpcAresRequest::OnAresBackupPollAlarm() {
+  {
+    grpc_core::MutexLock lock(&mu_);
+    ares_backup_poll_alarm_handle_.reset();
+    GRPC_ARES_WRAPPER_TRACE_LOG(
+        "request:%p OnAresBackupPollAlarm shutting_down=%d.", this,
+        shutting_down_);
+    if (!shutting_down_) {
+      for (auto it = fd_node_list_->begin(); it != fd_node_list_->end(); it++) {
+        if (!(*it)->already_shutdown) {
+          GRPC_ARES_WRAPPER_TRACE_LOG(
+              "request:%p OnAresBackupPollAlarm; ares_process_fd. fd=%s", this,
+              (*it)->polled_fd->GetName());
+          ares_socket_t as = (*it)->polled_fd->GetWrappedAresSocketLocked();
+          ares_process_fd(channel_, as, as);
+        }
+      }
+      if (!shutting_down_) {
+        EventEngine::Duration next_ares_backup_poll_alarm_duration =
+            calculate_next_ares_backup_poll_alarm_duration();
+        Ref(DEBUG_LOCATION, "OnAresBackupPollAlarm").release();
+        ares_backup_poll_alarm_handle_ = event_engine_->RunAfter(
+            next_ares_backup_poll_alarm_duration, [this] {
+              grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+              grpc_core::ExecCtx exec_ctx;
+              OnAresBackupPollAlarm();
```

Comment:
If there is no chance that the user callbacks will ever be called directly from a call stack descending from `OnAresBackupPollAlarm`, then the contexts would be unnecessary. On first read, that seems to be the case. Could be worth verifying and removing if this callback will only ever call wrapper-internal code.

Reply:
Excellent analysis. I think it's correct.

> is it worth to guarantee that every code path from grpc_ares_wrapper/EventEngine::DNSResolver out to its caller have an ExecCtx on the stack?

Per our discussion, every path that calls a user callback must have both ExecCtx types on the stack.